Thea: Fisherman Model

Gauthamastro opened the pull request:
## Describe your changes
1. Updates Thea to Fisherman model
2. Removes the Thea aggregator dependency
3. Automatic claim of Thea deposits
4. Bug fixes to Thea authority set rotation
5. On-chain signature aggregation
6. Ethereum and other EVM chains support

## Checklist before requesting a review
- [x] I have performed a self-review of my code.
- [x] If it is a core feature, I have added thorough tests.
- [x] I removed all Clippy and Formatting Warnings. 
- [x] I added required Copyrights.

Gauthamastro committed:
Add on-chain signature aggregation

Gauthamastro committed:
Added Fisherman and dealing with misbehaviour logic

Gauthamastro committed:
Thea core logic completed

Gauthamastro committed:
Fix runtime

zktony committed:
Added chnages for eth

Gauthamastro committed:
Update thea logic

Gauthamastro committed:
Added changes for Ethereum  (#896)

## Describe your changes
Changed types and added new extrinsic to support Eth.

Gauthamastro committed:
cargo fmt

Gauthamastro committed:
fix thea tests

Gauthamastro committed:
fix thea-executor tests

Gauthamastro committed:
Update Thea message handler

Gauthamastro committed:
Fix tests

Gauthamastro committed:
Fix benchmarks

Gauthamastro committed:
Resolve clippy errors

Gauthamastro committed:
cargo fmt

Gauthamastro committed:
Fix benchmarking

zktony committed:
Added sign_prehash

zktony committed:
Removed comment

Gauthamastro committed:
Changed sign to sign_prehash (#897)

## Describe your changes
Changed sign to sign_prehash.

Gauthamastro committed:
fix cargo test

zktony committed:
Added benchmarks

zktony committed:
Added on_init benchmark

zktony committed:
Added benchmark code for Thea Pallet and thea Executor Pallet (#905)

## Describe your changes
Added benchmark code for Thea Pallet and thea Executor Pallet.

serhii-temchenko committed:
feat(weights): Updated thea weights

zktony committed:
Resolved all the comments

zktony committed:
Resolved all the comments (#906)

## Describe your changes
Resolved all the comments.

zktony committed:
Fixed ci issues

zktony committed:
Fix CI Issues

serhii-temchenko committed:
feat(weights): Updated weights for thea related pallets

zktony committed:
Fixed tests

zktony committed:
Merge branch 'gj/thea-final' of https://github.com/Polkadex-Substrate/Polkadex into gj/thea-final

zktony committed:
Fixed tests

zktony committed:
Fixed CI

Gauthamastro committed:
Update the correct runtime spec_version

Gauthamastro committed:
Update the correct parachain runtime spec_version

serhii-temchenko committed:
feat(weights): Updated thea related pallets weights

Gauthamastro committed:
Thea: Fisherman and Batch Signing (#894)

## Describe your changes

This PR raises the final changes to core Thea logic by adding Fisherman
model for deposits and batch signing for withdrawals

Gauthamastro committed:
Update the versions

Gauthamastro committed:
Hot fixes

Gauthamastro committed:
Merge branch 'mainnet-release' into Develop

Gauthamastro committed:
Make the threshold check same in both mainnet and parachain

Gauthamastro committed:
Revert "Make the threshold check same in both mainnet and parachain"

This reverts commit 47113bbea82cee3b359530e6905318cc49b64db7.

Gauthamastro committed:
Change threshold check

Gauthamastro committed:
Change threshold check

Gauthamastro committed:
Change threshold check

Gauthamastro committed:
add priority

Gauthamastro committed:
Add send thea message extrinsic

Gauthamastro committed:
Added fixtures

Gauthamastro committed:
Update signing in tests

Gauthamastro committed:
cargo fmt

zktony committed:
Added benchmark

Gauthamastro committed:
Cargo fmt

Gauthamastro committed:
Cargo fmt

zktony committed:
Merge branch 'gj/thea-hot-fixes' of https://github.com/Polkadex-Substrate/Polkadex into gj/thea-hot-fixes

serhii-temchenko committed:
feat(weights): Updated thea message handler pallet weights

zktony committed:
Added send_thea_message benchmark.

zktony committed:
Updated benchmark code

Gauthamastro committed:
Update parachain runtime version

Gauthamastro committed:
Add allowlisted relayer

Gauthamastro committed:
Increment mainnet runtime

Gauthamastro committed:
Hot fixes for thea deposit

Gauthamastro committed:
Add check for next nonce

Gauthamastro committed:
Cargo fmt

Gauthamastro committed:
Gj/thea hot fixes (#908)

## Describe your changes

Hot fixes for parachain runtime upgrade

nuel77 commented on pallets/thea-executor/src/lib.rs:
```diff
@@ -535,16 +576,13 @@ pub mod pallet {
 			let deposit_amount = deposit.amount_in_native_decimals(metadata); // Convert the decimals configured in metadata
 
-			if !frame_system::Pallet::<T>::account_exists(recipient) {
+			if !frame_system::Pallet::<T>::account_exists(&deposit.recipient) {
```
If the account doesn't exist, can't we add this to a pallet account which is guaranteed to exist? this way, if any user makes the mistake of calling it without ed, the funds can still be given back to the user via governance if needed. it should be better than making the funds "vanish"

Gauthamastro replied:
@zktony What do you think of sending it to pallet account ? We may be need an governance-origin extrinsic to move the funds out but that's okay.

zktony replied:
yeah .. we can do that

Gauthamastro replied:
Resolved.

nuel77 commented on pallets/thea-executor/src/lib.rs:
```diff
@@ -520,9 +562,8 @@ pub mod pallet {
-				});
+				// Execute Deposit
+				Self::execute_deposit(deposit.clone())?;
 				Self::deposit_event(Event::<T>::DepositApproved(
```
ApprovedEvent is getting emitted after the claim event. This is breaking the indexer as approved event happened before claim before.

nuel77 commented on pallets/thea-executor/src/lib.rs:
```diff
@@ -520,9 +562,8 @@ pub mod pallet {
-				});
+				// Execute Deposit
+				Self::execute_deposit(deposit.clone())?;
 				Self::deposit_event(Event::<T>::DepositApproved(
```
ApprovedEvent is emitted after the claim event (within Self::execute_deposit). This is breaking indexers as the approved event happened before the claim in the previous code. If claim event is obsolete now, better to remove it.

Gauthamastro replied:
Claim event is there to handle pending deposits before upgrade if any, if that's not there, we can delete it now. In Fisherman model, claim process is not there anymore.

Gauthamastro replied:
As discussed with @nuel77, we are going to keep it for sometime.

Gauthamastro committed:
If account doesn't exist and swap fails, then deposit to system account and exit

Gauthamastro committed:
Increment runtime version

Gauthamastro committed:
Keep alive - true

Gauthamastro committed:
Add deposit failed event

Gauthamastro committed:
Add withdrawal failed event

Gauthamastro committed:
Add withdrawal failed event

serhii-temchenko commented:
Please let me know when this PR will be ready for review so I could update weights again after recent [changes](https://github.com/Polkadex-Substrate/Polkadex/pull/907/commits/dcd8da688f40874d2eee44e5a8ce271e1190ef4d) in benchmarks and thea-executor pallet.

zktony committed:
Added uncomp key handler

zktony committed:
Handled parse_compressed unwrap.

zktony committed:
Emit event

zktony committed:
Added address type

serhii-temchenko committed:
feat(thea): Updated weights

Commit:
stop auto upgrade

Commit:
stop auto upgrade

Gauthamastro committed:
Unsigned extrinsics for exchanges

Gauthamastro committed:
cargo fmt + stable

Gauthamastro committed:
Remove unstable features

zktony committed:
Added new RPC

zktony committed:
Updated toolchain version

zktony committed:
Updated toolchain

zktony committed:
Update to stable

zktony committed:
Fixed Cargo.lock

Gauthamastro committed:
cargo fmt

Gauthamastro committed:
delete rust toolchain

Gauthamastro committed:
remove unstable features

Gauthamastro committed:
Merge remote-tracking branch 'origin/zk-uncomp-key' into zk-uncomp-key

Gauthamastro committed:
cargo fmt

zktony committed:
Fixed tests

zktony committed:
Fixed formatting

zktony committed:
Merge branch 'zk-uncomp-key' of https://github.com/Polkadex-Substrate/Polkadex into zk-uncomp-key

zktony committed:
Updated param type

Gauthamastro committed:
Added support for Uncompressed Key (#910)

Gauthamastro committed:
Merge branch 'Develop' into gj/exchange-script

zktony committed:
Fixed clippy warnings

Gauthamastro committed:
Cargo fmt with stable

Gauthamastro committed:
Crowdloan Exchange Scripts (#912)

## Describe your changes

1. Added two unsigned extrinsics to initialize and claim rewards by
exchanges

## Issue ticket number and link
Resolves #911
https://github.com/Polkadex-Substrate/issue-tracker/issues/271
## Checklist before requesting a review
- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [x] I removed all Clippy and Formatting Warnings. 
- [ ] I added required Copyrights.

zktony committed:
Fixed empty ecdsa key

Gauthamastro committed:
Fixed empty ecdsa key (#914)

Gauthamastro committed:
cargo fmt

Gauthamastro committed:
Add failure events for thea

Gauthamastro committed:
cargo fmt

Gauthamastro committed:
Increment spec version

Gauthamastro committed:
Merge remote-tracking branch 'origin/Develop' into Develop

Gauthamastro committed:
fix compilation errors

Gauthamastro committed:
Merge remote-tracking branch 'origin/Develop' into Develop

Gauthamastro committed:
Remove authority check in deposit

Gauthamastro committed:
Increment parachain runtime version

zktony committed:
Fixed key

Gauthamastro committed:
zk-hotfix-key-fix (#915)

## Describe your changes
Fixed hashing of public key.

Gauthamastro committed:
Increment spec version
